src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java

```diff
@@ -0,0 +1,559 @@
+      if (proceeding.getAuditTrail() != null
+          && (latestKeyChange == null || latestKeyChange
+              .before(proceeding.getAuditTrail().getLastSaved()))) {
+        latestKeyChange = proceeding.getAuditTrail().getLastSaved();
+      }
+
+      if (proceeding.getScopeLimitations() != null && !proceeding.getScopeLimitations().isEmpty()) {
+        for (final ScopeLimitation scopeLimitation : proceeding.getScopeLimitations()) {
+          if (scopeLimitation.getAuditTrail() != null
+              && (latestKeyChange == null || latestKeyChange
+                  .before(scopeLimitation.getAuditTrail().getLastSaved()))) {
+            latestKeyChange = scopeLimitation.getAuditTrail().getLastSaved();
+          }
+        }
+      }
+
+    }
+
+    for (final Opponent opponent : application.getOpponents()) {
+      if (opponent.getAuditTrail() != null
+          && (latestKeyChange == null || latestKeyChange
+              .before(opponent.getAuditTrail().getLastSaved()))) {
+        latestKeyChange = opponent.getAuditTrail().getLastSaved();
+      }
+    }
+    return latestKeyChange;
+
+  }
+
+  /**
+   * Determines the unique mapping ID for an object, which could be either a Proceeding or an
+   * Opponent. If the object's EBS ID is null, constructs a new ID using the appropriate prefix and
+   * object's ID. If EBS ID is not null, it returns the EBS ID directly.
+   *
+   * @param opponentOrProceeding the object to determine the mapping ID for, which could be an
+   *                             instance of Proceeding or Opponent.
+   * @return the mapping ID or null if the object type is neither Proceeding nor Opponent
+   */
+  private String getOpaInstanceMappingId(final Object opponentOrProceeding) {
+    if (opponentOrProceeding instanceof final Proceeding proceeding) {
+      if (proceeding.getEbsId() == null) {
+        return InstanceMappingPrefix.PROCEEDING.getPrefix() + proceeding.getId();
+      }
+      return proceeding.getEbsId();
+    } else if (opponentOrProceeding instanceof final Opponent opponent) {
+      if (opponent.getEbsId() == null) {
+        return InstanceMappingPrefix.OPPONENT.getPrefix() + opponent.getId();
+      }
+      return opponent.getEbsId();
+    }
+    return null;
+  }
+
+  /**
+   * Retrieves the scope limitation identifier for a given proceeding. If multiple scope limitations
+   * exist, returns "MULTIPLE". Otherwise, returns the identifier of the single scope limitation.
+   *
+   * @param proceeding the proceeding to evaluate for scope limitations
+   * @return the scope limitation identifier or "MULTIPLE"
+   */
+  private String getRequestedScopeForAssessmentInput(final Proceeding proceeding) {
+    String scopeLimitations = null;
+    if (proceeding.getScopeLimitations().size() > 1) {
+      scopeLimitations = "MULTIPLE";
+    } else {
+      for (final ScopeLimitation scopeLimitation : proceeding.getScopeLimitations()) {
```

Comment:
This does a for loop just to return the first element in a list.

Reply:
yes, but i have amended the function into a lambda.

private String getRequestedScopeForAssessmentInput(final Proceeding proceeding) {
    return Optional.ofNullable(proceeding.getScopeLimitations())
        .filter(scopeLimitations -> !scopeLimitations.isEmpty())
        .map(scopeLimitations -> scopeLimitations.size() > 1 ? "MULTIPLE" :
            scopeLimitations.get(0).getScopeLimitation().getId())
        .orElse(null);
  }